app/controllers/reports/regions_controller.rb

```diff
@@ -134,29 +132,30 @@ def set_force_cache
   end
 
   def find_region
-    if current_admin.feature_enabled?("region_reports")
-      find_region_v2
-    else
-      find_region_v1
-    end
-  end
-
-  def find_region_v1
-    if report_params[:report_scope] == "facility_district"
-      @region = FacilityDistrict.new(name: report_params[:id], scope: current_admin.accessible_facilities(:view_reports))
-    else
-      slug = report_params[:id]
-      klass = region_class.classify.constantize
-      @region = klass.find_by!(slug: slug)
-    end
-  end
-
-  def find_region_v2
-    region_type = report_params[:report_scope]
-    @region = if region_type == "facility_district"
-      FacilityDistrict.new(name: report_params[:id], scope: current_admin.accessible_facilities(:view_reports))
+    region_source = authorize {
+      case region_class
+      when "FacilityDistrict"
+        scope = current_admin.accessible_facilities(:view_reports)
+        FacilityDistrict.new(name: report_params[:id], scope: scope)
+      when "FacilityGroup"
+        current_admin.accessible_facility_groups(:view_reports).find_by!(slug: report_params[:id])
+      when "Block" # we don't have first class auth on Blocks yet, so we authorize via the parent FacilityGroup
+        block = Region.find_by!(slug: report_params[:id], region_type: "block")
+        owning_facility_group = block.district_region.source
+        unless current_admin.accessible_facility_groups(:view_reports).include?(owning_facility_group)
+          raise UserAccess::NotAuthorizedError
+        end
+        block
+      when "Facility"
+        current_admin.accessible_facilities(:view_reports).find_by!(slug: params[:id])
+      else
+        raise "bad region_class #{region_class}"
+      end
+    }
+    @region = if current_admin.feature_enabled?(:region_reports)
+      region_source.region
     else
-      Region.where(region_type: region_type).find_by!(slug: report_params[:id])
+      region_source
```

Comment:
and for legacy reports, we want the source, so `Facility`, `FacilityGroup`, etc...